Fix uncertain mounts

gnufied opened the pull request:
Add uncertain state for volume mounts which will handle cleanup of CSI mounts correctly.

Fixes https://github.com/kubernetes/kubernetes/issues/82426 and https://github.com/kubernetes/kubernetes/issues/82340 and https://github.com/kubernetes/kubernetes/issues/83129

This does not fixes similar problem associated with raw block devices.  We are blocked on https://github.com/kubernetes/kubernetes/pull/74026 until we can fix raw block devices.


This PR fixes several issues by tracking certain `MountDevice` and `SetUp` failures as uncertain and adding them to actualStateofWorld so as volume can be cleaned up correctly when Pod is no longer scheduled on the node.  Previously these failures would cause volume to be not added to actualStateofWorld and hence volume may actually get mounted successfully but never get cleaned up properly. 

It does so by returning errors with additional context. Return value of `SetUp` and `MountDevice` functions now return following types of errors:

```
//   - TransientOperationFailure - operation failed with a transient error and could succeed if retried
//   - UncertainProgressError - operation failed but was started by reconciler and hence may be in-progress
//   - Error of any other type should be considered a final error and cause volume to not added to ASOW
```

```release-note
Track mount operations as uncertain if operation fails with non-final error
```

gnufied committed:
Add code for introducing uncertain state of mounts

Add a comment about volumestate

gnufied committed:
Add code to mark volume as uncertain

Update bazel files
Add tests for volume mounts in uncertain state

gnufied commented:
/sig storage

cc @kubernetes/sig-storage-bugs

gnufied commented:
/kind bug
/priority important-soon

jingxu97 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -355,6 +374,35 @@ type VolumeToMount struct {
+	VolumeMountUncertain VolumeMountState = "VolumeMountUncertain"
+
+	// VolumeNotMounted means volume has not been mounted in pod's local path
+	VolumeNotMounted VolumeMountState = "VolumeNotMounted"
```
do we really need this state?
Normally we can just use a boolean value to indicate the volume is already mounted if it is true. If it is false, that means we are not sure it is mounted or not, but still add it into the state and reconciler will update the state after it successfully mounted.

jingxu97 replied:
Never mind. Actually I see the point of having three states.

gnufied replied:
We can manage with boolean value of mount state but in a lot of places we typically have to carry a comment to explain a boolean value (/*MountUncertain*/) and stuff like that. I thought it might be best to explicit and let the field document what it is and that way - it will be more clear. What do you think?

gnufied replied:
Another thing I am thinking is - if mount failed with timeout first time and on second attempt mount failed with non-timeout(final) error then the volume should be moved to `VolumeNotMounted` state. I will update the code to do that.

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -127,6 +127,10 @@ type ActualStateOfWorld interface {
 	// actual state of the world.
 	GetMountedVolumes() []MountedVolume
 
+	// GetAllMountedVolumes returns list of all mounted volumes including
```
all "possibly" mounted volumes

gnufied replied:
fixed

gnufied commented:
/retest

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -354,12 +351,22 @@ func (asw *actualStateOfWorld) MarkVolumeAsUnmounted(
+}
+
+func (asw *actualStateOfWorld) MarkVolumeMountAsUncertain(markVolumeOpts operationexecutor.MarkVolumeMountedOpts) error {
+	markVolumeOpts.VolumeMountState = operationexecutor.VolumeMountUncertain
```
This line https://github.com/kubernetes/kubernetes/pull/82492/files#diff-450e811a4953f760ff1594ede8b2037eR753 already set the state to uncertain. Why need to set it up here too?

gnufied replied:
My thinking is - this is to safeguard against callers who may not be setting this correctly. The function is called `MarkVolumeMountAsUncertain`, so caller may expect that volume will be marked uncertain without explicitly specifying the state.

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -182,7 +194,9 @@ func (rc *reconciler) reconcile() {
+func (rc *reconciler) mountAttachVolumes() {
 	// Ensure volumes that should be attached/mounted are attached/mounted.
 	for _, volumeToMount := range rc.desiredStateOfWorld.GetVolumesToMount() {
 		volMounted, devicePath, err := rc.actualStateOfWorld.PodExistsInVolume(volumeToMount.PodName, volumeToMount.VolumeName)
```
PodExistsInVolume() returns true for MountUncertain volumes too?

gnufied replied:
good catch! fixed and added test.

jingxu97 commented:
As issue #82426 mentioned, if staging operation timeout, and pod is deleted, with this PR, reconciler will trigger unstage operation. Will it cause any problem if stage and unstage are proceeding at the same time?

gnufied commented:
@jingxu97 don't we have code in-place to prevent concurrent operations on a volume?

Madhu-1 commented on pkg/volume/csi/csi_client.go:
```diff
@@ -947,3 +978,27 @@ func (c *csiDriverClient) nodeGetVolumeStatsV1(
+	switch st.Code() {
+	case codes.Canceled, // gRPC: Client Application cancelled the request
+		codes.DeadlineExceeded,  // gRPC: Timeout
+		codes.Unavailable,       // gRPC: Server shutting down, TCP connection broken - previous CreateVolume() may be still in progress.
```
I think need to update comments.

gnufied replied:
fixe

gnufied committed:
Make sure PodExistsInVolume does not uses uncertain volumes

gnufied commented:
/retest

gnufied commented:
@jingxu97 addressed review comments PTAL

jingxu97 commented:
> a volume?

But if operation timesout, it will continue to trigger another operation?

gnufied commented:
@jingxu97 I think CSI spec is not very clear on what happens if `NodeStage` and `NodeUnstage` is called concurrently and I have filed - https://github.com/container-storage-interface/spec/issues/389 to seek some clarification.

humblec commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -272,13 +286,14 @@ func (rc *reconciler) reconcile() {
 	}
+}
 
-	// Ensure devices that should be detached/unmounted are detached/unmounted.
```
do you meant to comment about this function ?

gnufied replied:
I moved this to the caller.

gnufied commented:
we should be okay with running concurrent NodeUnstage operation while NodeStage operation is in progress. it is allowed as per CSI spec.  I reviewed code for ceph, gce and ebs drivers and found that only GCE driver implements this correctly. Filed issues against EBS and Ceph drivers - https://github.com/ceph/ceph-csi/issues/615 and https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/370

jsafrane reviewed (approved):
I tested it with a mock driver, it definitely helps, but I see some uncovered corner cases, see below.

gnufied commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -324,7 +325,8 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 	defer func() {
-		if err != nil {
+		// Only for non-timedout errors remove the mount directory
+		if err != nil && !volumetypes.IsOperationTimeOutError(err) {
```
It is possible to fix but this, it is kinda super ugly...

One can for example - return two errors from `MountDevice` call. One for CSI error and another for non-csi error. Both errors get recorded somewhere(asow or dsow not sure yet) and when next `MountDevice` call is made then both errors are supplied as arguments along with other args. So we can compare last csi error and current csi error and determine if operation is pending and not remove the directory for non-csi errors (and if both current and last csi-error is timeout).

It is no silver bullet however, I am afraid this will lead to orphan pod/volume directories left on the nodes and we have to carefully code this. There might be other unforeseen corner cases lurking around which we haven't thought through if we take this route.

Right now - my 2 cents are on calling current fix "good enough" and see if we receive bug reports from people about staging directories disappearing. I am not sure added complexity of tracking additional states is worth the cost. :(

jsafrane replied:
In external-provisioner we solved it by returning three values from [`ProvisionExt()`](https://github.com/jsafrane/sig-storage-lib-external-provisioner/blob/master/controller/volume.go#L76):

- `pv` - unrelated to NodeStage/Publish
- `err`: the error itself, mostly for logging
- `ProvisioningState` - state of the provisioning (when `err != nil`)
  - `Background` - CSI provisioner has reached CSI driver and got non-final error that means the provisioning may be in progress (e.g. timeout)
  - `Finished` - CSI provisioner has reached CSI driver and got final error that means the provisioning is for sure not in progress
  - `NoChange` - the provisioner did not reach the CSI driver (e.g. failed to get Secrets from API server). CSI provisioner has no clue in what state the provisioning is, but it is in the same state as before the call.

Caller of `ProvisionExt` (i.e. the provisioning lib) then keeps a list of provisionings in progress. IMO it's equal to uncertain state introduced in this PR.

An additional return val in `SetUp` / `MountDevice` ( / `TearDown` / `UnmountDevice`) could be enough, where the caller (`*Operation`) would mark the volume as uncertain when it receives `Background`, mark it as unmounted after `Finished` and leave the volume as it was before after `NoChange`.

gnufied replied:
fixed

gnufied commented on pkg/volume/util/types/types.go:
```diff
@@ -51,6 +51,31 @@ func (o *GeneratedOperations) Run() (eventErr, detailedErr error) {
 }
 
+// OperationTimedOutError indicates a particular volume operation has timed out.
+type OperationTimedOutError struct {
```
It is one wrapper that all errors non-final errors. Do we need different ones?

jsafrane replied:
No, I just don't like the name - it covers more than `OperationTimedOutError` name suggests

gnufied commented:
Yeah I think we might want to handle "uncertain" mount volume state (i.e `NodeUnpublish` timeouts) in a better way and mark volume mount as "uncertain" if unmount volume timesout.

gnufied committed:
Implement return status codes

gnufied committed:
Add code to handle Setup With Status tracking

gnufied committed:
Update generated files

gnufied committed:
Add tests for verifying in-progress state

gnufied committed:
Refactor NodeStage function

Timeout operations should result in
Fix unit tests for uncertainDeviceGlobalMounts

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -518,12 +535,26 @@ func (c *csiDriverClient) NodeStageVolume(ctx context.Context,
+		err = c.nodeStageVolumeV0(ctx, volID, publishContext, stagingTargetPath, fsType, accessMode, secrets, volumeContext, mountOptions)
+	}
+	// if there is no error and noStage was called return nil
+	if err == nil && nodeStageCalled {
```
fixd

gnufied commented on pkg/volume/csi/csi_client.go:
```diff
@@ -305,8 +313,17 @@ func (c *csiDriverClient) NodePublishVolume(
+		return nil
+	}
 
+	if err != nil {
```
fixed

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconciler_test.go:
```diff
@@ -1144,6 +1144,271 @@ func Test_Run_Positive_VolumeFSResizeControllerAttachEnabled(t *testing.T) {
+func waitForUncertainGlobalMount(t *testing.T, volumeName v1.UniqueVolumeName, asw cache.ActualStateOfWorld) {
+	// check if volume is globally mounted in uncertain state
+	err := retryWithExponentialBackOff(
+		time.Duration(500*time.Millisecond),
```
Since this is test code, should it be shorter (~100ms) and without exp. backoff? Or is the reconciler supposed to take long time?

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconciler_test.go:
```diff
@@ -1144,6 +1144,271 @@ func Test_Run_Positive_VolumeFSResizeControllerAttachEnabled(t *testing.T) {
+func waitForUncertainPodMount(t *testing.T, volumeName v1.UniqueVolumeName, asw cache.ActualStateOfWorld) {
+	// check if volume is locally pod mounted in uncertain state
+	err := retryWithExponentialBackOff(
+		time.Duration(500*time.Millisecond),
```
Since this is test code, should it be shorter (~100ms) and without exp. backoff? Or is the reconciler supposed to take long time?

(a `const` would be helpful)

jsafrane commented on pkg/kubelet/volumemanager/reconciler/reconciler_test.go:
```diff
@@ -1168,9 +1433,23 @@ func waitForMount(
 
+func waitForVolumeToExistInASW(t *testing.T, volumeName v1.UniqueVolumeName, asw cache.ActualStateOfWorld) {
+	err := retryWithExponentialBackOff(
+		time.Duration(500*time.Millisecond),
```
yeah, use `const` please

jsafrane reviewed (commented):
Initial review. I'll test it with a mock driver.

jsafrane commented:
/retest
Everything except kubernetes-verify are unrelated flakes

jsafrane commented:
Tested these scenarios:

---

1. Create a pod
2. NodeStage times out

What happened:
NodeStage was periodically called
Global dir was not cleaned after each timeout (it was before this PR)

---
1. Create a pod
2. NodeStage errors

What happened:
NodeStage was periodically called
Global dir was cleaned after each error
*Local pod directory was not cleaned up

---

1. Create a pod
2. NodeStage times out
3. Delete the pod

What happened:
NodeUnstage was called (this was expected)
*Local pod directory was not cleaned up

---

1. Create a pod
2. NodeStage errors
3. Delete the pod

What happened:
NodeUnstage was **not** called (this was expected)
*Local pod directory was not cleaned up

---

1. Create a pod
2. NodeStage times out
3. Delete the pod
4. NodeUnstage errors or times out

What happened:
NodeUnstage was periodically called (both "errors" and "times out" have the same result)

---

1. Create a pod
2. NodeStage times out
3. Delete the pod
4. NodeUnstage errors or times out
5. Create the same pod

What happened:
NodeStage was called (and NodePublish and pod got running)

---

In many tests, local pod directory was not cleaned either after NodeStage failure or NodeUnstage success. JSON remained there forever. Why was it created when
 NodePublish / SetUp was not called yet?


I'll continue with reconstruction tomorrow.

gnufied commented:
The reason local pod data is created even though Setup has not been called is - https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_plugin.go#L422

So the act of just instantiating the `Mounter` interface results in creation of the local pod directory and json file. However, if `MountDevice` failed for some reason - there is nothing in the code that will clean up that directory and data file. That is why - we have those files left. I also verified this behaviour with *before* this change and that also results in leftover local pod volume directory and data file.  So, it is at least not a regression but I think worth fixing.

gnufied committed:
Change signature of MountDevice function and remove MountDeviceWithStatusTracking

jsafrane commented:
> I have logged #84878 for fixing that behaviour.

Ack, we can fix it as a separate PR.

jsafrane commented:
Tested reconstruction, it works well when a volume is fully set up (i.e. NodePublish finished).

If kubelet is restarted when NodePublish has not finished yet (e.g. it has timed out), reconstruction checks that the pod directory is not a mount point and just cleans up the pod directory. This is a known limitation of reconstruction process - it considers only mount points for reconstruction. See https://github.com/kubernetes/kubernetes/issues/79980

gnufied committed:
Change interface of SetUp function

gnufied commented on pkg/volume/volume.go:
```diff
@@ -240,6 +246,9 @@ type DeviceMounter interface {
 	MountDevice(spec *Spec, devicePath string, deviceMountPath string) error
+
+	// MountDeviceWithStatusTracking is same as MountDevice except status of mount operation is also returned
+	MountDeviceWithStatusTracking(spec *Spec, devicePath string, deviceMountPath string) (volumetypes.OperationStatus, error)
```
Okay, so I ended up removing both of `*WithStatusTracking` functions and using plain `MountDevice` and `SetUp`.  I did not like the idea of introducing new interface because logically it did not fit in with existing interfaces.  

This does mean - surface area of the change will be bigger.

gnufied commented:
/retest

gnufied commented:
/retest

jsafrane commented:
/retest

jsafrane commented:
@jingxu97 @msau42 @davidz627, any comments? It looks fine from my POV.

jsafrane commented:
/retest

jsafrane commented:
/milestone 1.17

jsafrane commented:
@saad-ali, can you please add milestone? Thanks!

gnufied commented:
/test pull-kubernetes-e2e-gce-100-performance

gnufied commented:
/test pull-kubernetes-kubemark-e2e-gce-big

gnufied commented:
/retest

gnufied commented:
/retest

gnufied commented:
/retest

jingxu97 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -301,6 +309,12 @@ type mountedPod struct {
+	// volumeMountStateForPod stores state of volume mount for the pod. if it is:
+	//   - VolumeMounted: means volume for pod has been successfully mounted
+	//   - VolumeMountUncertain: means volume for pod may not be mounted, but it must be unmounted
+	//   - VolumeNotMounted: means volume for pod has not been mounted
```
do you need VolumeNotMounted state? I think the only action for this state is to delete the volume from the cache. This state is not used.

gnufied replied:
I think we can drop it.

jingxu97 commented on pkg/volume/util/operationexecutor/operation_executor.go:
```diff
@@ -161,18 +161,37 @@ func NewOperationExecutor(
 	}
 }
 
+// MarkVolumeMountedOpts is an struct to pass arguments to MountVolume functions
```
this struct to passed to different operations such as uncertain, unmounted etc, so how about call it MarkVolumeOpts?

gnufied replied:
ack. I will change it.

jingxu97 commented on pkg/volume/fc/attacher.go:
```diff
@@ -95,39 +96,42 @@ func (attacher *fcAttacher) GetDeviceMountPath(
+		if err != nil {
+			if os.IsNotExist(err) {
+				if err := os.MkdirAll(deviceMountPath, 0750); err != nil {
+					return err
```
it does not need to return OperationStatus?

gnufied replied:
It does return `OperationStatus` by calling `return  volumetypes.OperationFinished, mountInternal()` at the end. Some of this refactoring was hard to do in individual plugins because changing  return value of a volume interface function means modifying all those return statements in different places.  I have wrapped original `mountDevice` in internal `mountInternal` function and then we are returning both `OperationStatus` and `Error.

gnufied committed:
Rename MarkVolumeMountedOpts to MarkVolumeOpts

Also remove VolumeNotMounted state

jingxu97 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -98,33 +99,41 @@ func (c *csiMountMgr) CanMount() error {
 	csi, err := c.csiClientGetter.Get()
 	if err != nil {
-		return errors.New(log("mounter.SetUpAt failed to get CSI client: %v", err))
+		return opExitStatus, errors.New(log("mounter.SetUpAt failed to get CSI client: %v", err))
```
is it possible that temporarily failt to get csiClient, so that status should return NoChange?

gnufied replied:
may be yeah. I will change this code.

jingxu97 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -219,37 +220,39 @@ func (c *csiAttacher) GetDeviceMountPath(spec *volume.Spec) (string, error) {
 	if err != nil {
 		klog.Error(log("attacher.MountDevice failed while checking mount status for dir [%s]", deviceMountPath))
-		return err
+		return opExitStatus, err
```
is there some possible state that temporarily fail  to get mount state? so operationstatus is nochange?

gnufied replied:
I am not sure how this could happen and if this does - since we have not started `NodeStageVolume` call yet, it should be safe to mark the operation as "finished".  Are you thinking some corner case where `NodeStageVolume` was called but is in-progress and next `MountDevice` call fails in `isDirMounted` check and hence device is marked as not-mounted?

jingxu97 replied:
yes, that is the thought

jsafrane commented:
/retest

jsafrane commented:
/retest

(dbugging gce-big job, seems to fail too consistently in this PR)

jsafrane commented:
gce-big (https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/82492/pull-kubernetes-kubemark-e2e-gce-big/1194197582746750976/) fails because:

* Pod `test-uaaneu-3/small-deployment-150-6785544b7f-vs4n5` is scheduled to node `e2e-82492-ac87c-minion-group-89nk/kubelet-hollow-node-4tlsl`.

* The pod sets up Secrets just fine and starts successfully:

```
I1112 11:01:04.809803       6 kubelet.go:1947] SyncLoop (PLEG): "small-deployment-150-6785544b7f-vs4n5_test-uaaneu-3(6cd0a2a7-5a2f-415f-a165-8f4c3da17115)", event: &pleg.PodLifecycleEvent{ID:"6cd0a2a7-5a2f-415f-a165-8f4c3da17115", Type:"ContainerStarted", Data:"3e49643a9103c081"}
```

* clusterloader/v0.0.0 deletes the Secret used by the pod in API server:
```
I1112 11:03:24.643351       1 httplog.go:90] DELETE /api/v1/namespaces/test-uaaneu-3/secrets/small-deployment-150: (4.341218ms) 200 [clusterloader/v0.0.0 (linux/amd64) kubernetes/$Format 34.68.137.176:34018]
```

* The pod is deleted shortly after the secret was deleted:
    ```
    I1112 11:03:31.177689       1 httplog.go:90] DELETE /api/v1/namespaces/test-uaaneu-3/pods/small-deployment-150-6785544b7f-vs4n5: (5.408707ms) 200 [kube-controller-manager/v1.18.0 (linux/amd64) kubernetes/3222b3b/system:serviceaccount:kube-system:generic-garbage-collector [::1]:42968]
    ```
    And everything goes south from here.

    * kubelet still tries to SetUp the pod instead of TearDown after the pod was deleted:
        ```
        I1112 11:03:31.211101       6 kubelet.go:1918] SyncLoop (DELETE, "api"): "small-deployment-150-6785544b7f-vs4n5_test-uaaneu-3(6cd0a2a7-5a2f-415f-a165-8f4c3da17115)"
        I1112 11:03:31.211333       6 fake_cpu_manager.go:47] [fake cpumanager] RemoveContainer (container id: 165e512c66489898)
        I1112 11:03:31.211348       6 kuberuntime_container.go:591] Killing container "docker://165e512c66489898" with 2 second grace period
        I1112 11:03:31.638285       6 reconciler.go:267] operationExecutor.MountVolume started for volume "secret" (UniqueName: "kubernetes.io/secret/6cd0a2a7-5a2f-415f-a165-8f4c3da17115-secret") pod "small-deployment-150-6785544b7f-vs4n5" (UID: "6cd0a2a7-5a2f-415f-a165-8f4c3da17115")
        ```
    * And kubelet tries to SetUp the volume instead of TearDown forever. E.g. after 10 minutes:
        ```
        I1112 11:15:46.048244       6 reconciler.go:267] operationExecutor.MountVolume started for volume "secret" (UniqueName: "kubernetes.io/secret/6cd0a2a7-5a2f-415f-a165-8f4c3da17115-secret") pod "small-deployment-150-6785544b7f-vs4n5" (UID: "6cd0a2a7-5a2f-415f-a165-8f4c3da17115")
        E1112 11:15:46.048393       6 secret.go:200] Couldn't get secret test-uaaneu-3/small-deployment-150: secret "small-deployment-150" not found
        E1112 11:15:46.048485       6 nestedpendingoperations.go:270] Operation for "\"kubernetes.io/secret/6cd0a2a7-5a2f-415f-a165-8f4c3da17115-secret\" (\"6cd0a2a7-5a2f-415f-a165-8f4c3da17115\")" failed. No retries permitted until 2019-11-12 11:17:48.048454525 +0000 UTC m=+2195.419666335 (durationBeforeRetry 2m2s). Error: "MountVolume.SetUp failed for volume \"secret\" (UniqueName: \"kubernetes.io/secret/6cd0a2a7-5a2f-415f-a165-8f4c3da17115-secret\") pod \"small-deployment-150-6785544b7f-vs4n5\" (UID: \"6cd0a2a7-5a2f-415f-a165-8f4c3da17115\") : secret \"small-deployment-150\" not found"
        ```

So there actually may be a bug in this PR! VolumeManager should switch to TearDown instead of SetUp.

gnufied commented:
The problem appears to be caused by remounting mechanism whereas a failing remount can mark a previously successful volume as "unmounted".  I am fixing this and also adding additonal tests - https://github.com/kubernetes/kubernetes/pull/82492/files#diff-450e811a4953f760ff1594ede8b2037eR781

jsafrane reviewed (commented):
Found couple of nits while looking what went wrong with Secrets.

jsafrane commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -707,11 +707,23 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
+						klog.Errorf(volumeToMount.GenerateErrorDetailed("MountDevice.MarkDeviceAsUncertain failed", markDeviceUncertainError).Error())
+					}
+				case volumetypes.OperationFinished:
+					markDeviceUnmountError := actualStateOfWorld.MarkDeviceAsUnmounted(volumeToMount.VolumeName)
```
> The problem appears to be caused by remounting mechanism whereas a failing remount can mark a previously successful volume as "unmounted".

I think this is the place that broke Secrets in e2e tests.

Just an idea: with the new interface, Secrets, ConfigMaps, Projected and DownwardAPI volumes could return `NoChange` on Kube API errors and the volume status would not be changed.

gnufied replied:
We could but I am currently thinking of keeping that behaviour out of those drivers and just for CSI. The fix I am making does it such that - if a volume was mounted previously, then next mount(re) failure should not mark the volume as unmounted. 

Each case where `NoChange` or such new change is added - we will have to carefully review that code.

jsafrane commented:
I went trhough this PR again and it's mostly LGTM (except Secrets setup). If there are no objections and Secrets are fixed, I'd like to merge it into 1.17.

gnufied committed:
Handle the case of remounts correctly

jingxu97 commented on pkg/volume/csi/csi_client.go:
```diff
@@ -949,3 +971,27 @@ func (c *csiDriverClient) nodeGetVolumeStatsV1(
+		return false
+	}
+	switch st.Code() {
+	case codes.Canceled, // gRPC: Client Application cancelled the request
```
How about some code like PERMISSION_DENIED, UNAUTHENTICATED?

gnufied replied:
Elsewhere we have assumed those error codes to be "final" and unless CSI driver deployment is fixed, this error will keep happening.

jsafrane replied:
Since user is not authenticated or authorized we can safely assume that the operation did not even start. If we mark a volume as uncertain, we would most probably get PERMISSION_DENIED or UNAUTHENTICATED error during Unpublish / Unstage.

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -811,6 +814,47 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
+			}
+		}
+	case volumetypes.OperationFinished:
+		// Similarly only devices which were uncertain can be marked as unmounted
```
it might still have issue?
1. mount time out, mark as uncertain
2. second mount returns Finished due to some reason such as secret is deleted
3. mount succeeded?

jingxu97 replied:
I think we can simply do nothing (do not mark as mounted, nor mark it as unmounted) if volumetype is operationFinished for (re)mount

gnufied replied:
So basically same mount operation which was in-progress in step#1 failed in step#2 and then mount operation initiated in step#1 succeeds.   I agree that this could happen but:

1. for CSI drivers, they can't return a "final" error code for a in-progress operation and other code paths which use external API return `NoChange` status.
2. For secrets etc - I am going to make it to return `NoChange` on `remount` if Secret can not be found, as jan proposed in - https://github.com/kubernetes/kubernetes/pull/82492#discussion_r345795525 but since "uncertain" is never returned for secrets in general, the likelihood of this happening is low and will fixed for good in the follow up of secrets/configmaps PR.

jingxu97 replied:
Is there any bad consequence of leaving the volume as "uncertain"?

gnufied replied:
I can't think of any bad consequence(atm) except - I have tried to stick to existing data model used by volumemanager. In this model - if a volume is never successfully mounted, it is not recorded in ASOW. With above change you are proposing, known failed mounts will be kept in ASOW and should be unmounted/toredown when pod is deleted.

jingxu97 replied:
Agree your point. I just feel this way it might be safer and simpler. Otherwise, we have to make sure driver always return correct status just introduced. In previous code, successful mount will mark volume as mounted, failed mount does not unmark anything (of course, there is actually nothing to unmark). 

I think I am ok either way.

gnufied commented:
/retest

saad-ali commented:
/milestone v1.17

jingxu97 commented:
Do we need to handle unmount timeout case?
1. unmount timeout, but later succeed
2. pod is added back with volume, 
3. volume still in actual state, so no mount will be triggered?
If it is an issue, we can address in separate PR

gnufied commented:
@jingxu97 I think that is a valid point. Lets handle this as a separate issue - https://github.com/kubernetes/kubernetes/issues/85248

jsafrane commented:
/lgtm

gnufied commented:
/test pull-kubernetes-node-e2e-containerd

jsafrane commented:
I think containerd is broken also in other PRs and it's not Required.

jingxu97 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -287,25 +292,32 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
 				csiSource.NodeStageSecretRef.Namespace, csiSource.NodeStageSecretRef.Name, err)
-			return err
+			// if we failed to fetch secret then that could be a transient error
+			opExitStatus = volumetypes.OperationStateNoChange
```
if fail to get credentials, the state is OperationStateNoChange, I think it will not mark volume in the actual state. Then is there any code we will clean up metadata?

gnufied replied:
fixed

jingxu97 commented on pkg/volume/csi/csi_attacher.go:
```diff
@@ -262,10 +265,12 @@ func (c *csiAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMo
-		if err != nil {
+		// Only if there was an error and volume operation was considered
+		// finished, we should remove the directory.
+		if err != nil && opExitStatus == volumetypes.OperationFinished {
```
if fail to get credentials, the state is OperationStateNoChange, I think it will not mark volume in the actual state. Then is there any code we will clean up metadata?

gnufied replied:
fixed

jingxu97 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -199,26 +209,29 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 		nodePublishSecrets, err = getCredentialsFromSecret(c.k8s, secretRef)
 		if err != nil {
-			return fmt.Errorf("fetching NodePublishSecretRef %s/%s failed: %v",
+			opExitStatus = volumetypes.OperationStateNoChange
```
target_dir will not be cleaned up here?

gnufied replied:
As discussed previously with @jsafrane  - cleaning up of pod's local volume directory is broken regardless of what we do here. I was thinking of fixing this separately in - https://github.com/kubernetes/kubernetes/issues/84878

jingxu97 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -227,7 +240,8 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 	podAttrs, err := c.podAttributes()
 	if err != nil {
-		return errors.New(log("mounter.SetUpAt failed to assemble volume attributes: %v", err))
+		opExitStatus = volumetypes.OperationStateNoChange
```
target_dir will not be cleaned up here?

jingxu97 replied:
maybe move MkdirAll after getting these attributes?

gnufied replied:
same as above

gnufied committed:
Ensure that metadata directory is not created if secret is not found

To ensure that metadata directory is not created if secret is not
found, we will move fetching secrets bit more earlier in the code.

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -749,11 +750,23 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
 
 		// Execute mount
-		mountErr := volumeMounter.SetUp(volume.MounterArgs{
+		opExitStatus, mountErr := volumeMounter.SetUp(volume.MounterArgs{
```
Now we introduce this new operation state, should we log the returned state?

gnufied replied:
I am thinking that, while keeping all the underlying logic same, use special error types to indicate different state of failures. It will reduce the size of change and will allow us to keep volume plugin interface, more or less same. Thanks to @saad-ali for proposing this idea.

gnufied replied:
fixed

saad-ali commented on pkg/volume/volume.go:
```diff
@@ -239,7 +241,7 @@ type DeviceMounter interface {
 	// individual pods can then bind mount
 	// Note that devicePath can be empty if the volume plugin does not implement any of Attach and WaitForAttach methods.
-	MountDevice(spec *Spec, devicePath string, deviceMountPath string) error
+	MountDevice(spec *Spec, devicePath string, deviceMountPath string) (volumetypes.OperationStatus, error)
```
As far as I can tell the purpose of `volumetypes.OperationStatus` is to differentiate between 3 possible states: 1) success/no error (state certain), 2) error failed operation (state certain), and 3) error failed operation (uncertain state)?

There are two pieces of information here: 1) `error`/`no error`,  and 2) for `no error`, there are two possible sub-states: `certain`/`uncertain`.

Returning `error` AND an explicit `volumetypes.OperationStatus` seems wrong since it allows for the possibility of the two parameters to contradict each other.

I think typed errors would be cleaner here. But overall this is a nit.

gnufied replied:
fixed

gnufied committed:
Use typed errors for special casing volume progress

Use typed errors rather than operation status for
indicating operation progress

josiahbjorgaard commented:
/milestone v1.18

(code freeze v1.17)

gnufied commented:
/retest

davidz627 commented:
Please improve release note - right now it is not clear what is changing/affected.

gnufied commented:
@davidz627 done

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -274,13 +288,14 @@ func (rc *reconciler) reconcile() {
 		if !rc.desiredStateOfWorld.VolumeExists(attachedVolume.VolumeName) &&
 			!rc.operationExecutor.IsOperationPending(attachedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName) {
-			if attachedVolume.GloballyMounted {
+			if attachedVolume.DeviceMayBeMounted() {
```
not a big issue, but this function could be implemented locally?

gnufied replied:
I am thinking of leaving it as it is - since the resulting if statement looks quite big when placed inline and I do not like making a local helper function in this file.. :(

jingxu97 replied:
I mean you can have a local function to check device maybe mounted?

jingxu97 commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -274,13 +288,14 @@ func (rc *reconciler) reconcile() {
 	for _, attachedVolume := range rc.actualStateOfWorld.GetUnmountedVolumes() {
 		// Check IsOperationPending to avoid marking a volume as detached if it's in the process of mounting.
 		if !rc.desiredStateOfWorld.VolumeExists(attachedVolume.VolumeName) &&
 			!rc.operationExecutor.IsOperationPending(attachedVolume.VolumeName, nestedpendingoperations.EmptyUniquePodName) {
```
not your change, but I forgot why we need to check operationPending only for unmountDevice?

gnufied replied:
Here is the PR that implemented this - https://github.com/kubernetes/kubernetes/pull/40417

jingxu97 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -199,7 +205,8 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 			if err != nil {
-				return err
+				// we could have a transient error associated with fetching publish context
+				return volumetypes.NewTransientOperationFailure(err.Error())
```
why this one does not use fmt.Sprintf("") to generate errors as below?

gnufied replied:
I think mostly because this error here does not wrap existing error into a string. We can do that here though..

jingxu97 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -99,10 +100,14 @@ func (c *csiMountMgr) CanMount() error {
 }
 
 func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error {
+	return c.setupInternal(dir, mounterArgs)
```
why adding this setupInternal?

gnufied replied:
removed

jingxu97 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -272,18 +282,7 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
-			return fsGrpErr
-		}
-		return fsGrpErr
+		return volumetypes.NewUncertainProgressError(fmt.Sprintf("applyFSGroup failed for vol %s: %v", c.volumeID, err))
```
maybe add comment here. At this point, mount is successful, but we want to 
1. return error so that volume cannot be used since ownership is not set correctly
2. return NewUncertainProgressError so that volume will be marked in actual state and mount will be cleaned up when it is needed.

gnufied replied:
added comment

jingxu97 commented:
Overall looks good to me. A few small comments. Since this is a big change, maybe add more  design details in the PR comment section.

gnufied commented:
@jingxu97 added additional comments to the PR. ptal

jingxu97 commented on pkg/volume/util/types/types.go:
```diff
@@ -51,6 +51,57 @@ func (o *GeneratedOperations) Run() (eventErr, detailedErr error) {
+}
+
+// IsOperationFinishedError checks if given error is of type that indicates
+// operation is finished with an error.
```
with a FINAL error (not transient, or uncertain progress)

gnufied replied:
fixed

jingxu97 commented on pkg/volume/util/operationexecutor/operation_generator.go:
```diff
@@ -679,6 +683,49 @@ func (og *operationGenerator) GenerateMountVolumeFunc(
+	if volumetypes.IsOperationFinishedError(mountError) &&
+		actualStateOfWorld.GetDeviceMountState(volumeToMount.VolumeName) == DeviceMountUncertain {
+		// Only devices which were uncertain can be marked as unmounted
+		markDeviceUnmountError := actualStateOfWorld.MarkDeviceAsUnmounted(volumeToMount.VolumeName)
```
My thought is if you don't mark device as unmount when previously marked as uncertain. then you could simplify the logic. there is no need of TransientOperationError. MarkDeviceAsUnmounted will be triggered during volume tear down phase. 
What do you think?

gnufied replied:
There are two reasons, I am not very sure about taking this path:
1. volumemanager has been designed to not track volumes that are not mounted (or for which mount failed with 100% certainity). Now we are saying - we will keep volumes which have failed to mount in "uncertain" state and hence they will require "teardown" and "unmountdevice" calls.  This introduces additional design change into volumemanager and I hope that - no plugin breaks (or fail unmount call) when unmount is called for failed mounts.  My intention has been to stick to code paths which are known to work.

2. The second reason is - what are we supposed to return when transient error does happen in mount calls? 
    * Mount is called and kubelet fails to fetch the secret: - If we return `uncertainProgressError` for them, then volume will be added to ASOW in uncertain state, even though mount call has not even been attempted.  It seems wrong to accept a volume in ASOW. This PR also made a change that, fetching secret call has been moved up and hence if there was an error fetching secret, call returns early without even creating volume directories.
    * If we return a regular error for transient errors - then that would cause volume directories to be removed. This is a problem if, previous error was "UncertainProgressError". We do not want a transient error to remove volume directories, in case previous mount operation is still in-progress.

Given above reasons - IMO having a transientError is not problematic. It makes it easier to reason about state of volumemanager and mount failures.

gnufied replied:
Updated second case - since we moved secret fetching code to top of the function,  the defer is not executed if fetching secret fails.  But the main point IMO still stands.

jingxu97 replied:
> volumemanager has been designed to not track volumes that are not mounted (or for which mount failed with 100% certainity). Now we are saying - we will keep volumes which have failed to mount in "uncertain" state and hence they will require "teardown" and "unmountdevice" calls. This introduces additional design change into volumemanager and I hope that - no plugin breaks (or fail unmount call) when unmount is called for failed mounts. My intention has been to stick to code paths which are known to work.

For this one, actually plugin has to deal with "teardown while previous mount still in progress or actually failed mount already" after adding this uncertain state, I think. The sequence of event would be like
1. volume setup call returns with timeout error, add it into state
2. pod is deleted, volume tear down is called (meantime the mount might still in process or might failed or might succeeded)

> Mount is called and kubelet fails to fetch the secret(on first attempt): - If we return uncertainProgressError for them, then volume will be added to ASOW in uncertain state, even though mount call has not even been attempted. It seems wrong to accept a volume in ASOW. This PR also made a change that, fetching secret call has been moved up and hence if there was an error fetching secret, call returns early without even creating volume directories.

I thought mount is called after fetching the secret?

gnufied replied:
> I thought mount is called after fetching the secret?

I have been thinking about this and I think we still have a problem. Lets say - `MountDevice` is called and first attempt returns in `UncertainProgressError` which causes volume to be added to uncertain state. Second attempt however fails with a final error (not secret fetching error but lets say NodeStage itself failed) then because error was a final error - https://github.com/kubernetes/kubernetes/blob/6dbd521cfaf3b2abe69dce376b3b2a1cda5406e1/pkg/volume/csi/csi_attacher.go#L270 will remove the deviceMountPath from node. But volume is still in "Uncertain" state and must be cleaned up, but https://github.com/kubernetes/kubernetes/blob/6dbd521cfaf3b2abe69dce376b3b2a1cda5406e1/pkg/volume/csi/csi_attacher.go#L484 will throw an error because data directory from `deviceMountPath` is already removed..

gnufied replied:
What do we actually gain by removing `TransientOperationFailure` from this PR?

jingxu97 replied:
My thought was if both with and without transaction error can work correctly, I prefer without transaction for simplicity. Later we don't need to worry about later when other ppl work on this code, they need to understand this logic well and set correctly types of error.

But as you point it out, if mount fails, metadata will be cleaned. Then when reconciler calls unmount, it will fail to get metadata. The logic here can help remove volume from actual state cache, so unmount does not need to be triggered. 

So I think this is good. Just to make sure we add enough comment about different types of errors

jingxu97 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -254,10 +257,13 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 	if err != nil {
-		if removeMountDirErr := removeMountDir(c.plugin, dir); removeMountDirErr != nil {
-			klog.Error(log("mounter.SetupAt failed to remove mount dir after a NodePublish() error [%s]: %v", dir, removeMountDirErr))
+		// If operation finished with error then we can remove the mount directory.
```
MkdirAll() can be called right before NodePublishVolume call, right? so all previous errors does not need to worry about remove dir.

jsafrane commented on pkg/volume/csi/csi_client.go:
```diff
@@ -255,7 +259,10 @@ func (c *csiDriverClient) NodePublishVolume(
+	if err != nil && !isFinalError(err) {
+		return volumetypes.NewUncertainProgressError(err.Error())
+	}
+	return nil
```
`return err`!

It took me a while to find out why a pod is running after NodePublish error.

gnufied replied:
fixed. So we had tests for this but fake CSIClient had correct behaviour but real csi client had this bug.. :(

jsafrane reviewed (commented):
Testing with the mock driver and found these issues:

* User deletes a pod, `NodeUnstage` times out (it never succeeds), user re-creates the pod -> `NodePublish` is called. `NodeStage` was expected to fix uncertain Stage.

* User force-deletes a pod, `NodeUnpublish` times out (it never succeeds), user re-creates the pod -> `NodePublish` is called, followed by `NodeUnpublish`. ~The pod gets Running even though kubelet tries to `Unpublish` the volume.~
    ```
    gRPCCall: {"Method":"/csi.v1.Node/NodeUnpublishVolume",... "Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"}
    gRPCCall: {"Method":"/csi.v1.Node/NodeUnpublishVolume",... "Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"}
    gRPCCall: {"Method":"/csi.v1.Node/NodeUnpublishVolume",... "Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"}
    
    < pod created here >
    
    gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities",... "Error":""}
    gRPCCall: {"Method":"/csi.v1.Node/NodePublishVolume","Request":{"volume_id":"1","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/mypv/globalmount","target_path":"/var/lib/kubelet/pods/6c5f730f-ad66-4cdf-abec-c55a8c04b033/volumes/kubernetes.io~csi/mypv/mount","volume_capability":{"AccessType":{"Mount":{}},"access_mode":{"mode":1}},"volume_context":{"name":"Mock Volume 1","storage.kubernetes.io/csiProvisionerIdentity":"1532709836099-8081-csi-mock"}},"Response":{},"Error":""}
    gRPCCall: {"Method":"/csi.v1.Node/NodeUnpublishVolume","Request":{"volume_id":"1","target_path":"/var/lib/kubelet/pods/356b4361-7e05-47cc-beaa-e649363d1c5e/volumes/kubernetes.io~csi/mypv/mount"},"Response":null,"Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"}
    ```
    edit: my bad, this is correct. Kubelet published to a new pod dir and is still trying to unpublish the old pod dir.

gnufied committed:
Ensure that error is returned on NodePublish

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -442,20 +477,20 @@ func (asw *actualStateOfWorld) AddPodToVolume(
+			outerVolumeSpecName:    outerVolumeSpecName,
+			volumeGidValue:         volumeGidValue,
+			volumeSpec:             volumeSpec,
+			volumeMountStateForPod: markVolumeOpts.VolumeMountState,
```
fixed

gnufied commented:
@jsafrane I think fixing unmount code path we filed up as a follow up issue since this PR was getting too big as it is -  https://github.com/kubernetes/kubernetes/issues/85248

gnufied commented:
/retest

jsafrane commented:
> I think fixing unmount code path we filed up as a follow up issue since this PR was getting too big as it is - #85248

#85248 mentions only NodeUnpublish timeout, not NodeUnstage. If I remember correctly, old implementation of this PR called NodeStage after NodeUnstage timeout or error && new pod appeared. See that last test in https://github.com/kubernetes/kubernetes/pull/82492#issuecomment-550396856

jingxu97 commented:
@jsafrane In your test case, you mentioned local pod volume dir is created even when MountDevice timesout. Is this still the case. Because from the logic of the code, if MountDevice returns errors, it should not move to MountVolume step?

gnufied commented:
@jingxu97 that certainly would still be the case - as I mentioned in bug  - https://github.com/kubernetes/kubernetes/issues/84878 , the fact that just instantiating the `mounter` for CSI results in creation of pod local volume directories and they are left over even if `MountDevice` errors out. This is a bug. It exists in current k8s and this PR does not fixes it and hence I filed a separate bug to fix it. 

@jsafrane Are you sure `NodeStage` is not being called if `Unstage` times out but `NodUnpublish` succeeds? The way I see it - nothing in code would prevent calling `NodeStage` if volume was unmounted via `NodeUnpublish` because `stage` and `mount` operations are done as part of same operation and code does not check if volume is marked as "global mounted" before calling `NodeStage`. Here is my tests:

Pod is deleted and NodeUnpublish is successful:
```
gRPCCall: {"Method":"/csi.v1.Node/NodeUnpublishVolume","Request":{"volume_id":"4","target_path":"/var/lib/kubelet/pods/67560ff7-aa33-4d08-bc45-e74ba188603e/volumes/kubernetes.io~csi/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/mount"},"Response":{},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeUnstageVolume","Request":{"volume_id":"4","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/globalmount"},"Response":null,"Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"
}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeUnstageVolume","Request":{"volume_id":"4","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/globalmount"},"Response":null,"Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"
}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeUnstageVolume","Request":{"volume_id":"4","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/globalmount"},"Response":null,"Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"
}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeUnstageVolume","Request":{"volume_id":"4","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/globalmount"},"Response":null,"Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"
}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeUnstageVolume","Request":{"volume_id":"4","staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/globalmount"},"Response":null,"Error":"rpc error: code = DeadlineExceeded desc = Mock timeout"
}
```


Pod is re-created on same node:

```
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeStageVolume","Request":{"volume_id":"4","publish_context":{"device":"/dev/mock","readonly":"false"},"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/globalmount","volume_capability":{"A
ccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"name":"pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1","storage.kubernetes.io/csiProvisionerIdentity":"1576265141256-8081-io.kubernetes.storage.mock"}},"Response":{},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodePublishVolume","Request":{"volume_id":"4","publish_context":{"device":"/dev/mock","readonly":"false"},"staging_target_path":"/var/lib/kubelet/plugins/kubernetes.io/csi/pv/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/globalmount","target_path":"/var/l
ib/kubelet/pods/9fc9e8bf-b47f-41d1-bf1e-1c278147f72d/volumes/kubernetes.io~csi/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/mount","volume_capability":{"AccessType":{"Mount":{"fs_type":"ext4"}},"access_mode":{"mode":1}},"volume_context":{"name":"pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1",
"storage.kubernetes.io/csiProvisionerIdentity":"1576265141256-8081-io.kubernetes.storage.mock"}},"Response":{},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetCapabilities","Request":{},"Response":{"capabilities":[{"Type":{"Rpc":{}}},{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":2}}}]},"Error":""}
gRPCCall: {"Method":"/csi.v1.Node/NodeGetVolumeStats","Request":{"volume_id":"4","volume_path":"/var/lib/kubelet/pods/9fc9e8bf-b47f-41d1-bf1e-1c278147f72d/volumes/kubernetes.io~csi/pvc-bafb07c3-6f11-4985-9903-6492e46f5dc1/mount"},"Response":{"usage":[{"total":1073741824,"unit":1}]},"
Error":""}
```

Having said that - the reason `NodeStage` is being called even if `NodeUnstage` is never successful is because mock driver does not create a mount point on `deviceMountPath` - if it did then https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_attacher.go#L235 line will cause `MountDevice` to exit without ever calling `NodeStage`. So this is still a bug IMO and should be fixed, because for certain CSI driver implementation this will cause `NodeStage` to be never called.

msau42 commented:
I think we have separate problem of assuming things are mount points. 1) IsLikelyNotMount doesn't work for bind mounts on the rootfs, so this is problematic for a lot of local-based csi drivers 2) csi spec technically says things should be mount points, but k8s has been pretty lax about this so we would probably start breaking a lot of drivers if we tried to enforce it

gnufied commented:
Can we safely call `NodeStage` even though `deviceMountPoint` is already a mount point? All CSI calls being idempotent - it should be safe to do that. We might be able to remove checks for mount point entirely before calling `NodeStage` RPC.

jingxu97 commented:
This PR is lgtm. Thanks a lot @gnufied for putting great effort on this!

jsafrane commented:
Filed #86360 for Unstage timeout handling

/lgtm
/approve

gnufied commented:
/retest
